File: packages/sync-protocol/tests/merge_canon.mjs
/**
 * Canonicalization + merge-safe HLC rewrite helpers for the TS merge law suite.
 * Kept separate from merge_laws.test.mjs so helpers can be imported without
 * registering node:test cases.
 */

/** Keep generated HLCs inside the resolver's 24h skew horizon. */
export const MERGE_SAFE_PHYSICAL_MAX = () => Date.now() + 60 * 60 * 1000; // +1h

/**
 * Canonical JSON for join-semilattice equality (not object identity):
 * deep key sort + sorted deviceIds + frictionLog ordered by capturedAt.
 */
export function canonicalizeState(state) {
  const sortedFriction = [...(state.frictionLog ?? [])].sort((a, b) =>
    a.capturedAt < b.capturedAt ? -1 : a.capturedAt > b.capturedAt ? 1 : 0,
  );
  const normalized = {
    ...state,
    deviceIds: [...(state.deviceIds ?? [])].sort(),
    frictionLog: sortedFriction,
  };
  return JSON.stringify(sortKeysDeep(normalized));
}

export function sortKeysDeep(value) {
  if (Array.isArray(value)) return value.map(sortKeysDeep);
  if (value && typeof value === "object") {
    const out = {};
    for (const key of Object.keys(value).sort()) {
      out[key] = sortKeysDeep(value[key]);
    }
    return out;
  }
  return value;
}

/** Rewrite every HLC string's physical component into the merge-safe window. */
export function makeMergeSafe(state, maxPhysical = MERGE_SAFE_PHYSICAL_MAX()) {
  const rewrite = (hlc) => {
    if (typeof hlc !== "string" || hlc.length < 22) return hlc;
    const physical = Number(hlc.slice(0, 15));
    if (!Number.isFinite(physical) || physical <= maxPhysical) return hlc;
    return `${String(maxPhysical).padStart(15, "0")}${hlc.slice(15)}`;
  };

  return {
    ...state,
    profile: {
      ...state.profile,
      updatedAt: rewrite(state.profile.updatedAt),
    },
    mastery: Object.fromEntries(
      Object.entries(state.mastery).map(([k, m]) => [
        k,
        { ...m, lastExercisedAt: rewrite(m.lastExercisedAt) },
      ]),
    ),
    frictionLog: state.frictionLog.map((s) => ({
      ...s,
      capturedAt: rewrite(s.capturedAt),
    })),
    stateVector: Object.fromEntries(
      Object.entries(state.stateVector).map(([k, v]) => [k, rewrite(v)]),
    ),
  };
}

/**
 * Left-fold pairwise merges over a replica array (identity = first replica).
 * @param {{ merge: (a: object, b: object) => { merged: object } }} resolver
 * @param {object[]} replicas — length ≥ 1, same subjectId
 */
export function foldMerge(resolver, replicas) {
  if (replicas.length === 0) {
    throw new Error("FOLD_MERGE_EMPTY: need at least one replica");
  }
  let acc = replicas[0];
  for (let i = 1; i < replicas.length; i++) {
    acc = resolver.merge(acc, replicas[i]).merged;
  }
  return acc;
}

/**
 * Apply the SyncResponse compactedSampleTimestamps handshake: edge may drop
 * those friction samples from its local log once the merged document is adopted.
 * @param {object} state
 * @param {string[]} compactedSampleTimestamps
 */
export function applyCompactionHandshake(state, compactedSampleTimestamps) {
  const drop = new Set(compactedSampleTimestamps);
  return {
    ...state,
    frictionLog: (state.frictionLog ?? []).filter((s) => !drop.has(s.capturedAt)),
  };
}

/**
 * Fold with mid-stream compaction handshake: after the first `splitAt` replicas
 * are merged, announce their friction timestamps as compacted and prune those
 * keys from the remaining replicas before continuing the fold.
 *
 * @param {{ merge: (a: object, b: object) => { merged: object } }} resolver
 * @param {object[]} replicas
 * @param {number} [splitAt]
 */
export function foldMergeWithCompactionHandshake(resolver, replicas, splitAt = 1) {
  if (replicas.length === 0) {
    throw new Error("FOLD_MERGE_EMPTY: need at least one replica");
  }
  const cut = Math.min(Math.max(1, splitAt), replicas.length - 1);
  let mid = foldMerge(resolver, replicas.slice(0, cut + 1));
  const compactedSampleTimestamps = mid.frictionLog.map((s) => s.capturedAt);
  const remaining = replicas
    .slice(cut + 1)
    .map((r) => applyCompactionHandshake(r, compactedSampleTimestamps));
  for (const next of remaining) {
    mid = resolver.merge(mid, next).merged;
  }
  return { merged: mid, compactedSampleTimestamps };
}

/**
 * Deterministic Fisher–Yates from a uint32 seed (avoids uniqueArray rejection
 * sampling on index permutations).
 */
export function permuteFromSeed(n, seed) {
  const arr = Array.from({ length: n }, (_, i) => i);
  let s = seed >>> 0;
  for (let i = n - 1; i > 0; i--) {
    s = (Math.imul(s, 1664525) + 1013904223) >>> 0;
    const j = s % (i + 1);
    const tmp = arr[i];
    arr[i] = arr[j];
    arr[j] = tmp;
  }
  return arr;
}

/** Apply an index permutation to a replica array. */
export function permuteReplicas(replicas, order) {
  return order.map((i) => replicas[i]);
}

File: packages/cloud-orchestrator/tests/merge_canon.py
"""Canonicalization + merge-safe helpers — Python twin of ``merge_canon.mjs``.

Used by law / convergence suites against ``merge_states``.
"""

from __future__ import annotations

import json
import time
from typing import Any

from sutra_orchestrator.contract_models import CognitiveState
from sutra_orchestrator.crdt_merge import merge_states


def merge_safe_physical_max() -> int:
    """Keep generated HLCs inside a generous skew horizon (+1h)."""
    return int(time.time() * 1000) + 60 * 60 * 1000


def sort_keys_deep(value: Any) -> Any:
    if isinstance(value, list):
        return [sort_keys_deep(v) for v in value]
    if isinstance(value, dict):
        return {k: sort_keys_deep(value[k]) for k in sorted(value)}
    return value


def canonicalize_state(state: CognitiveState | dict[str, Any]) -> str:
    """Canonical JSON for join-semilattice equality (not object identity)."""
    if isinstance(state, CognitiveState):
        data = state.model_dump(mode="json")
    else:
        data = dict(state)
    normalized = {
        **data,
        "deviceIds": sorted(list(data.get("deviceIds") or [])),
        "frictionLog": sorted(
            list(data.get("frictionLog") or []),
            key=lambda s: s["capturedAt"] if isinstance(s, dict) else s.capturedAt,
        ),
    }
    return json.dumps(sort_keys_deep(normalized), separators=(",", ":"), ensure_ascii=True)


def make_merge_safe(
    state: dict[str, Any], *, max_physical: int | None = None
) -> dict[str, Any]:
    """Rewrite every HLC string's physical component into the merge-safe window."""
    cap = max_physical if max_physical is not None else merge_safe_physical_max()

    def rewrite(hlc: str) -> str:
        if not isinstance(hlc, str) or len(hlc) < 22:
            return hlc
        try:
            physical = int(hlc[:15])
        except ValueError:
            return hlc
        if physical <= cap:
            return hlc
        return f"{cap:015d}{hlc[15:]}"

    mastery = {
        k: {**m, "lastExercisedAt": rewrite(m["lastExercisedAt"])}
        for k, m in (state.get("mastery") or {}).items()
    }
    return {
        **state,
        "profile": {
            **state["profile"],
            "updatedAt": rewrite(state["profile"]["updatedAt"]),
        },
        "mastery": mastery,
        "frictionLog": [
            {**s, "capturedAt": rewrite(s["capturedAt"])} for s in state.get("frictionLog") or []
        ],
        "stateVector": {
            k: rewrite(v) for k, v in (state.get("stateVector") or {}).items()
        },
    }


def as_state(value: CognitiveState | dict[str, Any]) -> CognitiveState:
    if isinstance(value, CognitiveState):
        return value
    return CognitiveState.model_validate(value)


def merge_pair(
    left: CognitiveState | dict[str, Any],
    right: CognitiveState | dict[str, Any],
) -> CognitiveState:
    merged, _ = merge_states(as_state(left), as_state(right))
    return merged


def fold_merge(replicas: list[CognitiveState | dict[str, Any]]) -> CognitiveState:
    if not replicas:
        raise ValueError("FOLD_MERGE_EMPTY: need at least one replica")
    acc = as_state(replicas[0])
    for nxt in replicas[1:]:
        acc = merge_pair(acc, nxt)
    return acc


def apply_compaction_handshake(
    state: CognitiveState | dict[str, Any],
    compacted_sample_timestamps: list[str],
) -> dict[str, Any]:
    drop = set(compacted_sample_timestamps)
    data = (
        state.model_dump(mode="json")
        if isinstance(state, CognitiveState)
        else dict(state)
    )
    return {
        **data,
        "frictionLog": [s for s in data.get("frictionLog") or [] if s["capturedAt"] not in drop],
    }


def fold_merge_with_compaction_handshake(
    replicas: list[CognitiveState | dict[str, Any]],
    *,
    split_at: int = 1,
) -> tuple[CognitiveState, list[str]]:
    if not replicas:
        raise ValueError("FOLD_MERGE_EMPTY: need at least one replica")
    cut = min(max(1, split_at), len(replicas) - 1)
    mid = fold_merge(replicas[: cut + 1])
    compacted = [s.capturedAt for s in mid.frictionLog]
    for nxt in replicas[cut + 1 :]:
        mid = merge_pair(mid, apply_compaction_handshake(nxt, compacted))
    return mid, compacted


def permute_from_seed(n: int, seed: int) -> list[int]:
    """Deterministic Fisher–Yates from a uint32 seed."""
    arr = list(range(n))
    s = seed & 0xFFFFFFFF
    for i in range(n - 1, 0, -1):
        s = (s * 1664525 + 1013904223) & 0xFFFFFFFF
        j = s % (i + 1)
        arr[i], arr[j] = arr[j], arr[i]
    return arr


def permute_replicas(
    replicas: list[Any], order: list[int]
) -> list[Any]:
    return [replicas[i] for i in order]
